## 1.Build a regression model.

### Here I have used Multiple Linear regression model for which I have loaded data from previous parts of the project.

I tried to find out relation between 'Empty slots' with other variables like 'Bikes Avalable', 'Slots', 'Ebikes' which later I can use to predict empty slot(target variable) on the basis of other variables (feature variables)

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
import requests
import pandas as pd
import os
import json
stations_df = pd.read_csv('../data/citybikes_stations.csv')
restaurants_df=pd.read_csv('../data/correct_yelp_restaurants.csv')

merged_df_r = pd.merge(stations_df, restaurants_df, on=('Latitude','Longitude'))


In [2]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score


# Select features and target variable
features = ['Bikes Available', 'Slots','Ebikes']
target = 'Empty Slots'


# Prepare features (X) and target variable (y)
X = merged_df_r[features]
y = merged_df_r[target]

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize the linear regression model
model = LinearRegression()

# Train the model on the training data
model.fit(X_train, y_train)

# Make predictions on the test data
predictions = model.predict(X_test)

# Evaluate the model
mse = mean_squared_error(y_test, predictions)
r2 = r2_score(y_test, predictions)

print(f'Mean Squared Error: {mse}')
print(f'R-squared: {r2}')
print(f'intercept:{model.intercept_}')
print(f'coefficients:{model.coef_}')

Mean Squared Error: 0.7154265436231191
R-squared: 0.9959194645539086
intercept:-0.6325863295735772
coefficients:[-0.94975819  1.00883898 -0.28349051]


## Provide model output and an interpretation of the results. 

In [ ]:
Mean Squared Error: 0.7154265436231191
R-squared: 0.9959194645539086


1. **Intercept**: The intercept, represented as -0.6325863295735772 indicates the baseline value of the dependent variable in the absence of any predictor variables i.e. if features = ['Bikes Available', 'Slots','Ebikes'] are all 0 then this intercept will be value of 'empty slot' which means you will not get empty slots.

2. **Coefficients**: 
   [-0.94975819  1.00883898 -0.28349051] for ['Bikes Available', 'Slots','Ebikes'] respectively 
   
- suggests that -ve coefficient of 'Bike Available' means if this parameter increases possibility of finding 'Empty Slot' is less which is obvious because if bike is available at a slot it means it is occupying a slot.
- positive coefficient with 'Slots' means if slot increases possibility of finding 'Empty Slot' increases
- -ve coefficient of 'EBikes' means if this parameter increases possibility of finding 'Empty Slot' is less which is obvious because if an ebike is available at a slot it means it is occupying a slot.

3. **Mean Squared Error (MSE)**: A lower MSE of .715 indicates that the model's predictions are closer to the actual values on average.

4. **R-squared (R²)**:  an R-squared value of approximately 0.9958 means that about 99.59% of the variance in the dependent variable is predictable from the independent variables in the model. A higher R-squared value indicates a better fit of the model to the data.



# Stretch

## How can you turn the regression model into a classification model?

Here I have created a classification model using a Decision Tree Classifier to predict the 'Empty Slots Class' based on the specified features = ['Rating', 'Reviews','Banking','Slots'] using test data from merged_df_r dataframe.

I have set values for 
low_threshold and medium_threshold to define the bins for categorizing 'Empty Slots' into classes: 'Unlikely', 'Likely', and 'Highly'.

I have used test_size=0.2 in function train_test_split to divide 80% data as training set and 20% data as testing set

In [3]:
low_threshold = 10  # Adjust this value based on your dataset
medium_threshold = 30  # Adjust this value based on your dataset

# Define the bins and labels for classes
bins = [-float("inf"), low_threshold, medium_threshold, float("inf")]
labels = ['Unlikely', 'Likely', 'Highly']

# Binning the 'Empty Slots' into classes
merged_df_r['Empty Slots Class'] = pd.cut(merged_df_r['Empty Slots'], bins=bins, labels=labels)

features = ['Rating', 'Reviews','Banking','Slots']
target = 'Empty Slots Class'  # The new target variable

from sklearn.model_selection import train_test_split   #this function is used to create training and testing datasets
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

# Prepare features (X) and target variable (y)
X = merged_df_r[features]
y = merged_df_r[target]

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Initialize the Decision Tree classifier
classifier = DecisionTreeClassifier(random_state=42)

# Train the classifier on the training data
classifier.fit(X_train, y_train)

# Make predictions on the test data
predictions = classifier.predict(X_test)

# Evaluate the classifier
accuracy = accuracy_score(y_test, predictions)
print(f'Accuracy: {accuracy}')




Accuracy: 0.9781659388646288


with the help of DecisionTreeClassifier Model and accuracy of .9781 suggets that model is performing well with given features 'Rating', 'Reviews','Banking','Slots'] i.e. it can predict Empty Slot Class with given features ['Rating', 'Reviews','Banking','Slots'] accurately. However there is a scope to improve the model by providing more training data captured at different time intervals.